Hudl.Mjolnir/Command/CommandInvoker.cs

```diff
@@ -0,0 +1,238 @@
+﻿using Hudl.Common.Extensions;
+using Hudl.Mjolnir.Breaker;
+using Hudl.Mjolnir.Bulkhead;
+using Hudl.Mjolnir.External;
+using log4net;
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Hudl.Mjolnir.Command
+{
+    public interface ICommandInvoker
+    {
+        Task<CommandResult<TResult>> InvokeAsync<TResult>(AsyncCommand<TResult> command, OnFailure failureAction, long? timeoutMillis = null);
+        CommandResult<TResult> Invoke<TResult>(SyncCommand<TResult> command, OnFailure failureAction, long? timeoutMillis = null);
+
+        // TODO possible alternate signatures:
+        // 
+        // Could forego the CommandResult wrapper, which would enforce a OnFailure.Throw
+        //   TResult InvokeAndUnwrapOrThrow<TResult>(...)
+    }
+
+    // TODO what do timeouts/cancellations look like in exceptions now? make sure we didn't revert that logging change
+
+    public class CommandInvoker : ICommandInvoker
+    {
+        private readonly IStats _stats;
+
+        // TODO kind of ugly, rework this. they're lightweight to construct, though, and
+        // callers shouldn't be repeatedly constructing invokers.
+        // TODO make sure callers know not to repeatedly construct invokers :)
+        private readonly IBulkheadInvoker _bulkheadInvoker = new BulkheadInvoker(new BreakerInvoker());
+
+        public CommandInvoker() : this(CommandContext.Stats)
+        {
+            _stats = CommandContext.Stats; // TODO any risk here? should we just DI this? possibly not.
+        }
+
+        internal CommandInvoker(IStats stats, IBulkheadInvoker bulkheadInvoker = null)
+        {
+            if (stats == null)
+            {
+                throw new ArgumentNullException("stats");
+            }
+
+            _stats = stats;
+            if (bulkheadInvoker != null)
+            {
+                _bulkheadInvoker = bulkheadInvoker;
+            }
+        }
+
+        public async Task<CommandResult<TResult>> InvokeAsync<TResult>(AsyncCommand<TResult> command, OnFailure failureAction, long? timeoutMillis = null)
+        {
+            // This doesn't adhere to the OnFailure action because it's a bug in the code
+            // and should always throw so people see it and fix it.
+            EnsureSingleInvoke(command);
+            
+            var log = LogManager.GetLogger("Hudl.Mjolnir.Command." + command.Name);
+            var timeout = command.GetActualTimeout(timeoutMillis);
+
+            var invokeStopwatch = Stopwatch.StartNew();
+            var executeStopwatch = Stopwatch.StartNew();
+            var status = CommandCompletionStatus.RanToCompletion;
+            
+            var cts = timeout.HasValue
```

Comment:
Where are you passing through the Command's CancellationToken? i.e. if one has been explicitly passed into the call.
This might be happening somewhere else, but I can't find it at quick glance.